datumaro/cli/util/project.py

```diff
@@ -22,17 +28,104 @@ def generate_next_file_name(basename, basedir='.', sep='.', ext=''):
 
     return generate_next_name(os.listdir(basedir), basename, sep, ext)
 
-def generate_next_name(names, basename, sep='.', suffix='', default=None):
-    pattern = re.compile(r'%s(?:%s(\d+))?%s' % \
-        tuple(map(re.escape, [basename, sep, suffix])))
-    matches = [match for match in (pattern.match(n) for n in names) if match]
+_dataset_revpath_regex = None
+_full_revpath_regex = None
+def parse_full_revpath(s: str, ctx_project: Optional[Project]):
+    if ctx_project:
+        env = ctx_project.env
+    else:
+        env = Environment()
+
+    def parse_dataset_pathspec(s: str):
+        global _dataset_revpath_regex
+        if not _dataset_revpath_regex:
+            _dataset_revpath_regex = re.compile(
+                "%s(:%s)?" % \
+                (
+                    r"(?P<dataset_path>[^:]+)",
+                    r"(?P<format>.+)"
+                )
```

Comment:
At this point there's not much reason to build the regex from multiple parts. I would suggest to use `re.VERBOSE` and a multiline string instead.